---
title: "TCGA-BRCA-metadata"
author: "M.Mert Özgür"
date: '2022-07-25'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## TCGA-BRCA Data Analysis and Metadata Compilation

The project described below focuses on the analysis of the breast cancer (BRCA) data available on the TCGA-BRCA project. To attain that, four different data sets included in the TCGA-BRCA project are downloaded. Generated data is processed and necessary analysis to achieve proper data clearing have been done. Each of the assessed and cleared data (BRCA RNASeq, BRCA miRNA-Seq, BRCA clinical subtype, BRCA patient data) are being compiled to generate a metadata  which can be utilized by TCGA E#######.

Download required packages.
```{r, eval = F}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("TCGAbiolinks")
BiocManager::install("dplyr")
BiocManager::install("SummarizedExperiment")
BiocManager::install("edgeR")
BiocManager::install("tidyverse")
library(TCGAbiolinks)
library(dplyr)
library(SummarizedExperiment)
library(edgeR)
library(tidyverse) #package attaches ggplot2.
```

Get information about the TCGA-BRCA project.
```{r}
getProjectSummary("TCGA-BRCA")
```

### Query-Download-Prepare the BRCA-RNASeq data
```{r eval= F}
query_RNAseq = GDCquery(
  project = "TCGA-BRCA",
  data.category = "Transcriptome Profiling",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  data.type = "Gene Expression Quantification"
)
colnames(query_RNAseq) # see queried RNASeq data information, Total of 12 column. 

GDCdownload(
  query = query_RNAseq,
  method = "client",
  directory = "TCGA_TranscriptomeProfiling_BRCA"
)
BRCA_TranscriptomeProfiling = GDCprepare(
  query = query_RNAseq,
  save = TRUE,
  save.filename = "prepdata_BRCA-RNAseq",
  directory = "TCGA_TranscriptomeProfiling_BRCA"
)
BRCA_RNASeqraw = assay(BRCA_TranscriptomeProfiling)
saveRDS(object = BRCA_RNASeqraw,
        file = "BRCA_RNASeqraw.RDS") #RNASeq raw data is saved as R object.
saveRDS(object = BRCA_TranscriptomeProfiling,
        file = "BRCA_TranscriptomeProfiling.RDS")
```

### Query-Download-Prepare BRCA-miRNA-Seq data
```{r eval= F}
query_miRSeq = GDCquery(
  project = "TCGA-BRCA",
  data.category = "Transcriptome Profiling",
  experimental.strategy = "miRNA-Seq",
  workflow.type = "BCGSC miRNA Profiling",
  data.type = "miRNA Expression Quantification"
)
colnames(query_miRSeq)
GDCdownload(
  query = query_miRSeq,
  method = "client",
  directory = "BRCA-miRNA-Seq"
)
BRCA_miRNASeqraw = GDCprepare(
  query = query_miRSeq,
  save = TRUE,
  save.filename = "prepdata_BRCA_miRNA-Seq",
  directory = "BRCA-miRNA-Seq"
)
saveRDS(object = BRCA_miRNASeqraw,
        file = "BRCA_miRNASeqraw.RDS") #miRNA-Seq raw data is saved as a R object. 

```

### Get Molecular Subtype data
```{r eval= F}
BRCA_subtype = TCGAquery_subtype("BRCA")
saveRDS(object = BRCA_subtype, file = "BRCA_subtype.RDS")
```

### Get patient clinical data
```{r eval= F}
BRCA_patient_clinical = as.data.frame(colData(BRCA_TranscriptomeProfiling))
saveRDS(object = BRCA_patient_clinical, file = "BRCA_patient_clinical.RDS")
```

### TCGA-BRCA RNASeq data analysis and processing
Gene version analysis for TCGM-BRCA RNASeq Data. Different versions of the genes having the same name (gene_name) whose Ensemble gene_ids differ are studied. Different versions are labeled with their Ensemble gene_ids. BRCA RNASeq data is prepared to study gene versions. (Versions of the genes:Genes having the same name but different ensemble gene_id due to various reasons)

```{r}
BRCA_TranscriptomeProfiling = readRDS("~/Desktop/TCGA-BRCA-metadata/TCGA-BRCA-metadata/BRCA_TranscriptomeProfiling.RDS")
brca_geneinfo = as.data.frame(rowData(BRCA_TranscriptomeProfiling))
brca_gene.version = brca_geneinfo %>% 
  select(gene_id, gene_name) %>% mutate(duplicated_info = duplicated(brca_geneinfo$gene_name) | duplicated(brca_geneinfo$gene_name, fromLast= T)) 

for(p in 1:length(brca_gene.version$gene_name)) {
  if (brca_gene.version$duplicated_info[p] == FALSE){
    
  } else {
    brca_gene.version[p,"gene_name"] = paste(brca_gene.version[p,"gene_name"],brca_gene.version[p,"gene_id"] ,sep = "_")
  }
}

brca_gene.version = select(brca_gene.version,-c("duplicated_info"))
saveRDS(object = brca_gene.version, file = "brca_gene.versionENSG.RDS") #save gene version data frame as a R object. 
```

gene_ids in the RNASeq data are replaced with the new form of the gene_name which includes gene version information. By doing that, different versions can be reached by their name with their associated Ensemble gene_id and can be analyzed for elimination, adjustment etc.
```{r}
BRCA_RNASeqraw = readRDS("~/Desktop/TCGA-BRCA-metadata/TCGA-BRCA-metadata/BRCA_RNASeqraw.RDS")
BRCA_RNASeqraw = as.data.frame(BRCA_RNASeqraw)
rownames(BRCA_RNASeqraw) = brca_gene.version[,"gene_name"] 
```

RNASeq data normalization is performed before the analysis of the gene versions to avoid data loss.
```{r}
BRCA_RNASeq = cpm(BRCA_RNASeqraw)
BRCA_RNASeq = log(BRCA_RNASeq + 1, base = 10) #Data turned to a class of matrix, convert it before go. 
BRCA_RNASeq= as.data.frame(BRCA_RNASeq)
```

Transpose the RNASeq data. Gene names are in columns now, thus, data cleaning based on genes is possible.
```{r}
BRCA_RNASeq = t(BRCA_RNASeq)  
BRCA_RNASeq = as.data.frame(BRCA_RNASeq) 
```

Merging the BRCA RNASeq and miRNA-Seq data will be done according to the sample_id. Thus sample_id (substring of the TCGA barcode) were extracted and mutated (appended) in a separate column named sample_id. Since the subtype data and clinical data will be merged via the patient_id, patient_id also included in the data.
```{r}
BRCA_RNASeq <- BRCA_RNASeq %>% 
  mutate(sample_id = substr(rownames(.), 1,16 ))%>%
  mutate(patient_id = substr(rownames(.),1,12)) %>%
  select(sample_id, patient_id, everything())
saveRDS(object = BRCA_RNASeq, file = "BRCA_RNASeq.RDS")  #BRCA RNASeq data which was prepared for cleaning process is ready. Data is saved.

```

### TCGA-BRCA miRNA-Seq data analysis and processing
```{r}
BRCA_miRNASeqraw = readRDS("~/Desktop/TCGA-BRCA-metadata/TCGA-BRCA-metadata/BRCA_miRNASeqraw.RDS")
miRNA = select(BRCA_miRNASeqraw, c(starts_with("miRNA_ID"),starts_with("read_count") ))
miRNA_names = select(miRNA, c(starts_with("miRNA_ID")))
miRNA_counts = select(miRNA, starts_with("read_count"))
```

miRNA-Seq data is normalized.Column naming (patient barcode) are adjusted in the data (unnecessary parts removed).
```{r}
miRNA_counts = cpm(miRNA_counts)
miRNA_counts = log(miRNA_counts + 1, base = 10)

colnames(miRNA_counts) = gsub("read_count_","",colnames(miRNA_counts))

miRNA_counts = t(miRNA_counts)#miRNA Seq data is transposed.
miRNA_counts = as.data.frame(miRNA_counts)
colnames(miRNA_counts) = miRNA_names$miRNA_ID #miRNA names are placed. 

```

sample_id and patient_id are appended in BRCA_miRNA data. The data is ready for merging.
```{r}
BRCA_miRNA = miRNA_counts %>% 
  mutate(sample_id = substr(rownames(.),1,16)) %>% 
  mutate(patient_id = substr(rownames(.),1,12)) %>% 
  select(sample_id,patient_id,everything())
saveRDS(object = BRCA_miRNA, file= "BRCA_miRNA.RDS") #BRCA miRNA-Seq data which was prepared for cleaning process is ready. Data is saved.
```

A variable named immune_signature is created and appended in the BRCA_RNASeq data.
```{r}
RNASeq_w_immune = BRCA_RNASeq %>% 
  mutate(immune_signature = (CD8A + CD3E + GZMB + PRF1) / 4)
saveRDS(object = RNASeq_w_immune, file= "RNASeq_w_immune.RDS")
```

Merging the RNASeq_w_immune and BRCA_miRNA data by sample_id.
```{r}
RNASeq_miRNASeq = full_join(RNASeq_w_immune,BRCA_miRNA, by = "sample_id")
saveRDS(object = RNASeq_miRNASeq, file= "RNASeq_miRNASeq.RDS")
```

### Clinical Data Analysis
There are two data sets which include clinical information about the samples. Accordingly, some of the variables in those two data sets are common for each sample and furthermore, either of them may include more recent information. Analysis workflow begins with the exploratory analysis of each variables in these two data sets. After identifying common variables, each of the duplicates (common variables-columns) are assessed in order to identify more recent information. To analyze missing data, function na.analyze is constructed as data structure necessitates.  


#### Function: na.analyze()
This function takes a data frame as an input. Since the data may contain missing data as  factor -"NA" or <NA> or both, na.analyze function enables to differentiate and count them separately. na.analyze() generates an output of data frame which shows: each column name and index of the input data frame, total number of "NA" count in each column and total number of <NA> in each column. Thus, the three variables of the output data frame are: column-names, NA_ascharacter_count and NA_asmissing_count.
```{r}
na.analyze = function(p) {
  NA_ascharacter = c()
  NA_asmissing = c()
  NA_analysis = as.data.frame(colnames(p))
  for (c in 1:nrow(NA_analysis)) {
    missing_counter = 0
    character_counter = 0 
    for (e in 1:nrow(p[c])) {
      missing_stat = is.na(p[e,c])
      missing_aschar_stat= p[e,c]
      if (missing_stat == TRUE) {
        next
      } else if (missing_aschar_stat == "NA") {
        character_counter = character_counter + 1
      }
    }
    NA_ascharacter = append(NA_ascharacter,character_counter)
    missing_counter = sum(is.na(p[c]))
    NA_asmissing = append(NA_asmissing,missing_counter)
  }
  
  NA_analysis = NA_analysis %>%  mutate(NA_ascharacter_count = NA_ascharacter) %>% mutate(NA_asmissing_count = NA_asmissing)
  return(NA_analysis)
}
```


```{r}
BRCA_patient_clinical = readRDS("~/Desktop/TCGA-BRCA-metadata/TCGA-BRCA-metadata/BRCA_patient_clinical.RDS")
BRCA_subtype = readRDS("~/Desktop/TCGA-BRCA-metadata/TCGA-BRCA-metadata/BRCA_subtype.RDS")

na.analyzed_patient_clinical = na.analyze(BRCA_patient_clinical)
na.analyzed_subtype = na.analyze(BRCA_subtype)


na.analyzed_patient_clinical = na.analyzed_patient_clinical %>% mutate(total_NA = NA_ascharacter_count + NA_asmissing_count) %>% 
  mutate(Available_from1226 = 1226- total_NA)
  
na.analyzed_subtype = na.analyzed_subtype %>% mutate(total_NA = NA_ascharacter_count + NA_asmissing_count) %>% 
  mutate(Available_from1087 = 1087- total_NA)

```

BRCA_subtype data consists of 24 variables. BRCA_patient_clinical data consists of 85 variables. According to analysis, all of the 24 variable in subtype data relocated on the last 24 column of the BRCA_patient_clinical data. There are also some common variables in the BRCA_patient_clinical data which have a duplicate within the same data frame. All of the duplicate variables were analyzed and variables including more recent information are selected before merging of the clinical data to the transcriptome profiling data. 

```{r}
withinpatient = na.analyzed_patient_clinical[62:85,]
common_comparison_table = cbind(withinpatient,na.analyzed_subtype)
View(common_comparison_table)
View(na.analyzed_patient_clinical)
```

Based on the results of the above missing data analysis, all of the BRCA_subtype data which are already included in BRCA_patient_clinical data are not updated. Hence, BRCA_patient_clinical versions of these variables (columns) contains more up-to-date information. Data clearing based on the results described is performed. 

```{r}
BRCA_patient_clinical = select(BRCA_patient_clinical,-c(paper_pathologic_stage,paper_days_to_last_followup,paper_Tumor_Grade,
                                                        paper_vital_status,paper_days_to_birth,paper_days_to_death))
saveRDS(object = BRCA_patient_clinical, file= "BRCA_patient_clinical_clear.RDS")
```
Merging the RNASeq_w_immune and BRCA_miRNA data by sample_id.

```{r}
BRCA_patient_clinical_clear <- readRDS("~/Desktop/TCGA-BRCA-metadata/TCGA-BRCA-metadata/BRCA_patient_clinical_clear.RDS")
#Those columns are also dropped in order to eliminate malfunctions which may occur during the merging. 
BRCA_patient_clinical_clear = select(BRCA_patient_clinical_clear,-c(bcr_patient_barcode,paper_patient,sample,sample_submitter_id,submitter_id))

RNASeq_miRNASeq =  RNASeq_miRNASeq %>%  rename_at("patient_id.x", ~"patient") %>% 
  select(-patient_id.y)
BRCA_metadata = full_join(RNASeq_miRNASeq,BRCA_patient_clinical_clear, by = "patient")
saveRDS(object = BRCA_metadata, file= "BRCA_metadata.RDS")
```


### Data Analysis
Scatter plot: hsa-mir-155 vs CD8A 
```{r}
BRCA_metadata %>% 
  ggplot(aes(`hsa-mir-155`,CD8A)) +
  geom_point(color = 'blue') +
  geom_smooth()
  
```
Scatter plot: CD3E vs CD8A 
```{r}
RNASeq_miRNASeq %>% 
  ggplot(aes(CD3E,CD8A)) +
  geom_point(color = 'orange') +
  geom_smooth(color = "blue" )

```
Scatter plot: immune_signature vs FOXP3
```{r}
RNASeq_miRNASeq %>% 
  ggplot(aes(immune_signature,FOXP3)) +
  geom_point(color = 'red') +
  geom_smooth(color = "green" )
```

CD8A in two gender classes (male vs female) (jitters)
```{r}
BRCA_metadata %>% 
  ggplot(aes(x=gender, y=CD8A, fill= gender)) +
  geom_boxplot()+
  geom_jitter(color="black",size=0.4, alpha= 0.9) +
  ggtitle("CD8A: male vs female")

```
Kaplan Meier curve: Immune signature categorized at median value as high and low

```{r}
install.packages("survminer")
install.packages("survival")
library(survminer)
library(survival)
BRCA_metadata <- readRDS("~/Desktop/TCGA-BRCA-metadata/TCGA-BRCA-metadata/BRCA_metadata.RDS")
brca_surv = select(BRCA_metadata,c(patient,days_to_last_follow_up,days_to_death,vital_status,
                                   immune_signature))

brca_surv = brca_surv[1:1548,] #last rows containing all-NAs are eliminated.
median_immune_signature = median(brca_surv$immune_signature, na.rm = TRUE)

brca_surv = brca_surv %>% 
  mutate(immune_category = case_when(
    immune_signature > median_immune_signature ~ "high" , 
    immune_signature < median_immune_signature ~ "low"
  )) %>% 
  mutate(days_to_event = case_when(
    is.na(days_to_death) == FALSE ~ days_to_death,
    TRUE ~ days_to_last_follow_up
  )) %>% 
  mutate(status = case_when(
    vital_status == "Alive" ~ 1,
    vital_status =="Dead" ~ 0
  ))

fit <- surv_fit(Surv(days_to_event, status) ~ immune_category, data = brca_surv)
ggsurvplot(fit,
           data=brca_surv,
           risk.table = TRUE,
           pval = TRUE,
           conf.int = TRUE,
           xlab = "Time in days",
           ggtheme = theme_bw() )

```



```{r}
surv_CD8A = select(BRCA_metadata[1:1548,],c(patient,days_to_last_follow_up,days_to_death,vital_status,
                                   CD8A))


median_CD8A = median(surv_CD8A$CD8A, na.rm = TRUE)

surv_CD8A = surv_CD8A %>% 
  mutate(CD8A_category = case_when(
    CD8A > median_CD8A ~ "high" , 
    CD8A < median_CD8A ~ "low"
  )) %>% 
  mutate(days_to_event = case_when(
    is.na(days_to_death) == FALSE ~ days_to_death,
    TRUE ~ days_to_last_follow_up
  )) %>% 
  mutate(status = case_when(
    vital_status == "Alive" ~ 1,
    vital_status =="Dead" ~ 0
  ))

fit_CD8A <- surv_fit(Surv(days_to_event, status) ~ CD8A_category, data = surv_CD8A)
ggsurvplot(fit_CD8A,
           data=surv_CD8A,
           risk.table = TRUE,
           pval = TRUE,
           conf.int = TRUE,
           xlab = "Time in days",
           ggtheme = theme_bw() )
```



